File: flut/flutter/services/text_editing.py
from typing import override

from flut._flut.engine import FlutValueObject, _flut_pack_value
from flut._flut.runtime import _flut_unpack_required_field
from flut.dart.ui import TextAffinity


class TextSelection(FlutValueObject):
    _flut_type = "TextSelection"

    def __init__(
        self,
        *,
        baseOffset: int,
        extentOffset: int,
        affinity: TextAffinity = TextAffinity.downstream,
        isDirectional: bool = False,
    ) -> None:
        super().__init__()
        self.baseOffset = baseOffset
        self.extentOffset = extentOffset
        self.affinity = affinity
        self.isDirectional = isDirectional

    @staticmethod
    def _flut_unpack(data: dict) -> "TextSelection":
        return TextSelection(
            baseOffset=_flut_unpack_required_field(data, "baseOffset"),
            extentOffset=_flut_unpack_required_field(data, "extentOffset"),
            affinity=_flut_unpack_required_field(data, "affinity"),
            isDirectional=_flut_unpack_required_field(data, "isDirectional"),
        )

    @override
    def _flut_pack(self) -> dict[str, object]:
        result = self._flut_base_props()
        result["baseOffset"] = _flut_pack_value(self.baseOffset)
        result["extentOffset"] = _flut_pack_value(self.extentOffset)
        result["affinity"] = _flut_pack_value(self.affinity)
        result["isDirectional"] = _flut_pack_value(self.isDirectional)
        return result

File: flut/.flutter/lib/flutter/services/text_editing.dart
import 'package:flutter/services.dart';
import 'package:flut/dart/ui.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/runtime.dart';

class FlutTextSelection extends FlutValueObject {
  final TextSelection selection;

  const FlutTextSelection(this.selection) : super('TextSelection');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    result['baseOffset'] = selection.baseOffset;
    result['extentOffset'] = selection.extentOffset;
    result['affinity'] = const FlutTextAffinity().flutEncode(
      selection.affinity,
    );
    result['isDirectional'] = selection.isDirectional;
    return result;
  }

  static TextSelection? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return TextSelection(
      baseOffset: runtime.unpackRequiredField<int>(data, 'baseOffset'),
      extentOffset: runtime.unpackRequiredField<int>(data, 'extentOffset'),
      affinity: runtime.unpackRequiredField<TextAffinity>(data, 'affinity'),
      isDirectional: runtime.unpackRequiredField<bool>(data, 'isDirectional'),
    );
  }
}
